finality-aleph/src/network.rs

```diff
@@ -141,148 +228,195 @@ enum SessionStatus {
         }
     }
 }
 
-impl<B: BlockT + 'static, N: Network<B> + Clone> ConsensusNetwork<B, N> {
+impl<D: Clone + Encode + Decode, B: BlockT + 'static, N: Network<B> + Clone>
+    ConsensusNetwork<D, B, N>
+{
     /// Create a new instance.
-    pub(crate) fn new(network: N, protocol: impl Into<Cow<'static, str>>) -> Self {
-        let protocol = protocol.into();
-        let (net_command_tx, net_command_rx) = mpsc::unbounded();
+    pub(crate) fn new(network: N, protocol: Cow<'static, str>) -> Self {
+        let (commands_for_session, commands_from_user) = mpsc::unbounded();
         ConsensusNetwork {
             network,
             protocol,
             sessions: Arc::new(Mutex::new(HashMap::new())),
-            net_command_tx,
-            net_command_rx,
+            commands_for_session,
+            commands_from_user,
             peers: Arc::new(Mutex::new(Peers::new())),
-            phantom: PhantomData,
+            _phantom: PhantomData,
         }
     }
 
-    pub(crate) fn session_manager(&self) -> SessionManager {
+    pub(crate) fn session_manager(&self) -> SessionManager<D> {
         SessionManager {
+            peer_id: self.network.peer_id(),
             sessions: self.sessions.clone(),
-            net_command_tx: self.net_command_tx.clone(),
+            commands_for_session: self.commands_for_session.clone(),
         }
     }
 
-    fn sample_random_peer(&self) -> Option<PeerId> {
-        self.peers.lock().sample_random()
+    fn send_message(&self, peer_id: &PeerId, message: InternalMessage<D>) {
+        self.network
+            .send_message(*peer_id, self.protocol.clone(), message.encode());
+    }
+
+    fn on_incoming_data(&self, session_id: SessionId, data: D) {
+        let mut sessions = self.sessions.lock();
+        if let Some(session_data) = sessions.get_mut(&session_id) {
+            if session_data.status == SessionStatus::InProgress {
+                if let Err(e) = session_data.data_for_user.unbounded_send(data) {
+                    //TODO: need to write some logic on when an session should be terminated and make sure
+                    // that there are no issues with synchronization when terminating.
+                    session_data.status = SessionStatus::Terminated;
+                    debug!(target: "afa", "Error {:?} when passing a message event to session {:?}.", e, session_id);
+                }
+            }
+        }
     }
 
-    fn send_message(&self, peer_id: PeerId, session_id: SessionId, message: &Vec<u8>) {
-        self.network.send_message(
+    fn on_incoming_authentication(&self, auth_data: AuthData, signature: Signature) {
+        let enc_auth_data = auth_data.encode();
+        let AuthData {
+            session_id,
             peer_id,
-            self.protocol.clone(),
-            (session_id, message).encode(),
-        );
+            node_id,
+        } = auth_data;
+        if let Some(session_data) = self.sessions.lock().get(&session_id) {
```

Comment:
There might be (?) this annoying issue that some node starts a session 1-2 secs later than others and many authentications for the new session get lost forever.  I would say -- leave a TODO for this, and one can work on this in a separate task, as the PR is quite large already.

Reply:
Authentications are sent not only when a new session starts, but also on every `peer_connected` event, so I think this should actually work. Although I'm not convinced I get the semantics of the `peer_connected` events quite right, so I'm not sure.

Reply:
The thing is though that the peer is already connected when a new session starts.

Reply:
Ah, I see what you mean, the session is not started so we don't iterate over it. Yeah, that is a serious problem – either keeping a forward set of sessions, or resending authentications in the way we do with some other messages would solve it, but both require significant work. I'll add a task.